### PR TITLE
Get rid of Zephyr's prefer-later-defaults behavior

### PR DESCRIPTION
**Pull request by ulfalizer**
This PR gets rid of Zephyr's prefer-later-defaults behavior for Kconfig defaults.

The commit message for `Kconfig: Use the first default with a satisfied condition` (below) gives some motivations.

Documentation updates are included as well, and `boilerplate.cmake` prints a note about the new behavior so that it's easier to discover. See the other commit messages.

```
Kconfig: Use the first default with a satisfied condition

Up until now, Zephyr has patched Kconfig to use the last 'default' with a
satisfied condition, instead of the first one. I'm not sure why the patch was
added (it predates Kconfiglib), but I suspect it's related to Kconfig.defconfig
files.

There are at least three problems with the patch:

  1. It's inconsistent with how Kconfig works in other projects, which might
     confuse newcomers.

  2. Due to oversights, earlier 'range' properties are still preferred, as well
     as earlier 'default' properties on choices.

     In addition to being inconsistent, this makes it impossible to override
     'range' properties and choice 'default' properties if the base definition
     of the symbol/choice already has 'range'/'default' properties.

     I've seen errors caused by the inconsistency, and I suspect there are
     more.

  3. A fork of Kconfiglib that adds the patch needs to be maintained.

Get rid of the patch and go back to standard Kconfig behavior, as follows:

  1. Include the Kconfig.defconfig files first instead of last in
     Kconfig.zephyr.

  2. Include boards/Kconfig and arch/<arch>/Kconfig first instead of last in
     arch/Kconfig.

  3. Include arch/<arch>/soc/*/Kconfig first instead of last in
     arch/<arch>/Kconfig.

  4. Swap a few other 'source's to preserve behavior for some scattered symbols
     with multiple definitions.

     Swap 'source's in some no-op cases too, where it might match the intent.

  5. Reverse the defaults on symbol definitions that have more than one
     default.

     Skip defaults that are mutually exclusive, e.g. where each default has an
     'if <some board>' condition. They are already safe.

  6. Remove the prefer-later-defaults patch from Kconfiglib.

Testing was done with a Python script that lists all Kconfig symbols/choices
with multiple defaults, along with a whitelist of fixed symbols. The script
also verifies that there are no "unreachable" defaults hidden by defaults
without conditions

As an additional test, zephyr/.config was generated before and after the change
for several samples and checked to be identical (after sorting).

This commit includes some default-related cleanups as well:

  - Simplify some symbol definitions, e.g. where a default has 'if FOO' when
    the symbol already has 'depends on FOO'.

  - Remove some redundant 'default ""' for string symbols. This is the implicit
    default.

Piggyback fixes for swapped ranges on BT_L2CAP_RX_MTU and
BT_L2CAP_TX_MTU (caused by confusing inconsistency).

Piggyback some fixes for style nits too, e.g. unindented help texts.
```

### CODE REVIEW

File: doc/porting/board_porting.rst
Comment (dbkinder) [inline doc/porting/board_porting.rst]: For (future) clarity, change to (verify the right version here):

    Versions of Zephyr prior to 1.13 used a custom ...

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/porting/board_porting.rst
Comment (ulfalizer) [reply doc/porting/board_porting.rst]: Not sure what to put here, as I expect it to go in before the 1.13 release. I'll stall a bit.

File: doc/porting/board_porting.rst
Comment (ulfalizer) [reply doc/porting/board_porting.rst]: Fixed

File: doc/porting/board_porting.rst
Comment (ulfalizer) [reply doc/porting/board_porting.rst]: Fixed

File: doc/porting/board_porting.rst
Comment (ulfalizer) [reply doc/porting/board_porting.rst]: Updated it to say "prior to early August 2018 (during development of Zephyr 1.13), Zephyr used a custom patch..."

Is that okay?

**Review by SebastianBoe — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]
Looks fine from a Bluetooth perspective

File: cmake/app/boilerplate.cmake
Comment (SebastianBoe) [reply cmake/app/boilerplate.cmake]: Fine by me.

**Review by jukkar — APPROVED** [review]
+1 to networking changes

File: cmake/app/boilerplate.cmake
Comment (ulfalizer) [reply cmake/app/boilerplate.cmake]: Done

**Comment by ulfalizer** [comment]
@dbkinder *ping*

**Review by dbkinder — APPROVED** [review]
works for me, thanks!

**Comment by ulfalizer** [comment]
@nashif @carlescufi @SebastianBoe 
Should be ready to merge now, unless you want to wait longer.

I sent a message to `zephyr-devel` yesterday.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@ulfalizer I am happy with this, but I'd like @nashif to approve before I merge, since this is an important change.

**Review by MaureenHelm — APPROVED** [review]

**Comment by ulfalizer** [comment]
Found some goof-ups with `arch/xtensa/soc/*/Kconfig` includes being nested inside the wrong menu in `arch/<arch>/Kconfig` files.

Fixed now.

**Comment by ulfalizer** [comment]
@nashif ping

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
any reason why this is marked as DNM?

**Comment by carlescufi** [comment]
@nashif no. This is a remnant of the initial RFC. Please merge

**Pull request merged**
